pkg/trace/api/telemetry.go

```diff
@@ -0,0 +1,136 @@
+// by simply parametrizing metric tags, and logger names
+type telemetryMultiTransport struct {
+	Transport http.RoundTripper
+	Endpoints []*config.Endpoint
+}
+
+func (m *telemetryMultiTransport) roundTrip(req *http.Request, endpoint *config.Endpoint) (*http.Response, error) {
+	tags := []string{
+		fmt.Sprintf("endpoint:%s", endpoint.Host),
+	}
+	defer func(now time.Time) {
+		metrics.Timing("datadog.trace_agent.telemetry_proxy.roundtrip_ms", time.Since(now), tags, 1)
+	}(time.Now())
+
+	req.Host = endpoint.Host
+	req.URL.Host = endpoint.Host
+	req.URL.Scheme = "https"
+	req.Header.Set("DD-API-KEY", endpoint.APIKey)
+
+	resp, err := m.Transport.RoundTrip(req)
+	if err != nil {
+		metrics.Count("datadog.trace_agent.telemetry_proxy.roundtrip.errors", 1, tags, 1)
+	}
+	return resp, err
+}
+
+// RoundTrip sends request first to MainTarget, then sends a copy of main request to every configurged
+// AdditionalTargets receiptient.
+//
+// All requests will be sent irregardless of any errors
+// If any request fails, the error will be logged. Only main target's
+// error will be propagated via return value
+func (m *telemetryMultiTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	if len(m.Endpoints) == 1 {
+		return m.roundTrip(req, m.Endpoints[0])
+	}
+	slurp, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return nil, err
+	}
+	newreq := req.Clone(req.Context())
+	newreq.Body = ioutil.NopCloser(bytes.NewReader(slurp))
+	// despite the number of endpoints, we always return the response of the first
+	rresp, rerr := m.roundTrip(newreq, m.Endpoints[0])
+	for _, endpoint := range m.Endpoints[1:] {
+		newreq := req.Clone(req.Context())
+		newreq.Body = ioutil.NopCloser(bytes.NewReader(slurp))
+		if resp, err := m.roundTrip(newreq, endpoint); err == nil {
+			// we discard responses for all subsequent requests
+			io.Copy(ioutil.Discard, resp.Body)
+			resp.Body.Close()
+		} else {
+			log.Error(err)
+		}
+	}
+	return rresp, rerr
+}
+
+// telemetryProxyHandler parses returns a new HTTP handler which will proxy requests to the configured intakes.
+// If the main intake URL can not be computed because of config, the returned handler will always
+// return http.StatusInternalServerError along with a clarification.
+func (r *HTTPReceiver) telemetryProxyHandler() http.Handler {
+	// extract and validate Hostnames from configured endpoints
+	var endpoints []*config.Endpoint
+	for _, endpoint := range r.conf.TelemetryConfig.Endpoints {
+		u, err := url.Parse(endpoint.Host)
+		if err != nil {
+			log.Errorf("Error parsing apm_config.telemetry endpoint: %s", endpoint.Host)
+			continue
+		}
+		if u.Host != "" {
+			endpoint.Host = u.Host
+		}
+
+		endpoints = append(endpoints, endpoint)
+	}
+	if len(endpoints) == 0 {
+		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			log.Errorf("Telemetry proxy couldn't forward a request - no valid endpoints found")
```

Comment:
This log needs to be outside of the handler. Otherwise it'll show every time someone accesses this. We want to see the error immediately when the handler is initialized.